state-synchronizer/src/coordinator.rs

```diff
@@ -189,27 +202,65 @@ impl<T: ExecutorProxyTrait> SyncCoordinator<T> {
         }
     }
 
-    async fn request_sync(&mut self, request: SyncRequest) {
-        self.known_version = self
+    /// Sync up coordinator state with the local storage.
+    async fn sync_state_with_local_storage(&mut self) {
+        self.local_state = self
             .executor_proxy
-            .get_latest_version()
+            .get_local_storage_state()
             .await
-            .expect("[state sync] failed to fetch latest version from storage");
+            .expect("[state sync] failed to sync with local storage");
+    }
+
+    /// The highest available version in the local storage (even if it's not covered by the LI).
+    fn highest_version_in_local_storage(&self) -> u64 {
+        std::cmp::max(
+            self.local_state.highest_local_li.ledger_info().version(),
+            self.local_state.highest_committed_version,
+        )
+    }
+
+    /// In case there has been another pending request it's going to be overridden.
+    /// The caller will be notified about request completion via request.callback oneshot:
+    /// at that moment it's guaranteed that the highest LI exposed by the storage is equal to the
+    /// target LI.
+    /// StateSynchronizer assumes that it's the only one modifying the storage (consensus is not
+    /// trying to commit transactions concurrently).
+    async fn request_sync(&mut self, request: SyncRequest) {
+        if self.role != RoleType::Validator {
+            error!("Sync requests are not expected for non-validator nodes");
+            return;
+        }
```

Comment:
I don't know much about these code and whether it's feasible or not, but intuitively I'll extract common functions and have a FullNodeSynchronizer and ValidatorSynchronizer instead of holding a role to differentiate them

Reply:
We should look at it together. Overall, the differences are small and might not be worth the separations (at least I don't see an elegant way to separate it to a separate trait right now). The only difference is that the Validator role has the concept of a "sync request", which is checked upon commit processing during chunk response.